### ISSUE TITLE
[Epic] Multi-agent communication channels

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/thrust#266]
## Epic: Multi-agent communication channels
Roadmap Phase 2 final unchecked item. Confirmed not started. Needs an API design pass before implementation — filing as an epic proposal.

## Why an epic
"Communication channels" spans several distinct capabilities (explicit message-passing actions, shared/broadcast observations, learned communication protocols). Scope must be pinned down by design before any Builder work.

## Open design questions
- Explicit discrete/continuous message actions appended to the action space, vs. a separate comms channel in the env API?
- Does this extend the `MultiAgentEnvironment` trait, or live in a new adapter (cf. the Bucket Brigade Slepian-Wolf adapter)?
- Differentiable (learned) comms vs. fixed protocol?
- `crossbeam-channel` is already an optional dep — relevant for the runtime plumbing.

## Candidate phase breakdown (to be refined by Architect)
1. Design note: comms model + trait/API changes.
2. Env-side: message channel in the multi-agent env contract + a reference env.
3. Policy-side: consume/emit messages; train end-to-end.

## Acceptance criteria (epic-level)
- [ ] Design note / ADR approved.
- [ ] Decomposed into implementation issues.

**Comment by rjwalters** [comment]
## Curator Guidance

Curating this issue per user request (epic sweep). The deliverable scoped here is the **design note + child-issue decomposition**, not the comms implementation itself. The curator guidance below reflects that framing.

---

## Technical Verification

The following was verified against `origin/main` before writing this guidance.

### `MultiAgentEnvironment` trait (`src/multi_agent/environment.rs`)

The shipped trait is Burn-agnostic and clean. Current surface:

```rust
pub trait MultiAgentEnvironment: Environment {
    fn num_agents(&self) -> usize;
    fn get_agent_observation(&self, agent_id: usize) -> Vec<f32>;
    fn agent_action_space(&self, agent_id: usize) -> Vec<usize>;
    fn step_multi(&mut self, actions: &[Vec<i64>]) -> MultiAgentResult;
    fn active_agents(&self) -> Vec<bool>;
}
```

`MultiAgentResult` carries per-agent `observations`, `rewards`, `terminated`, `truncated`, and a shared `info: HashMap<String,String>`. There is no first-class message slot. Agent-to-agent comms today must either be embedded into the observation/action space or smuggled through `info`.

### Bucket Brigade "Slepian-Wolf adapter"

The issue conflates two distinct things; both exist, but neither is a general comms framework:

1. **BB env signal action dim** (`src/env/games/bucket_brigade/env.rs`): The action vector is `[house_index, mode, signal]` (length 3). The `signal` dim (0=REST, 1=WORK) is effectively a 1-bit broadcast message; received signals from all agents appear as `round1_signals` in each agent's observation vector. This is "embedded comms" — baked into action/obs spaces for one specific env.

2. **`JointMultiAgentTrainer`** (`src/multi_agent/joint.rs`): The "Slepian-Wolf adapter" referenced in the issue doc is this joint trainer, used for the cross-agent representational redundancy penalty `L_red = λ * Σ_{i<j} || corr(Z_i, Z_j) ||_F²`. This is a **training coupling** mechanism, not a runtime message-passing API.

The BB env demonstrates that action-space extension works for fixed binary vocab, but it requires env-specific observation layout changes and doesn't generalize.

### `src/multi_agent/messages.rs`

Already contains `Experience`, `PolicyUpdate`, `TrainingStats`, `ControlMessage`. These are **training-infrastructure** channels (simulator thread → learner thread). They do not address agent-to-agent comms during rollout. No `AgentMessage` type exists.

### `crossbeam-channel` (`Cargo.toml:99`)

```toml
crossbeam-channel = { version = "0.5", optional = true }
# enabled by the "training" feature
```

Already present as an optional dep. Relevant for the learner/simulator infrastructure, but for synchronous `step_multi` agent-to-agent comms **no new channel usage is needed** — agents share the environment's mutable state within a single step call. Channels become relevant only for async deferred-delivery comms (Phase 3+ scope).

---

## Answering the Four Design Questions

These answers should anchor the design note.

**Q1: Action-space extension vs. separate comms channel in env API?**

Recommend **action-space extension as the default model**: messages are an additional group of action dims on the sender, and received messages appear as additional observation dims on the receiver. This is:
- Already proven by the BB env
- Gym/Gymnasium-compatible (no new trait methods needed for simple cases)
- Trainable end-to-end without framework changes (multi-discrete policy already supported)

For the general case (variable vocab, typed messages, partial delivery), add a thin optional `CommunicatingEnvironment` supertrait.

**Q2: Extend `MultiAgentEnvironment` vs. new adapter?**

Recommend **new supertrait** `CommunicatingEnvironment: MultiAgentEnvironment`:
- Extending the base trait is a breaking change for all existing implementors (snake, matching pennies, bucket brigade, joint env)
- A supertrait with default impls (`fn message_obs_size(&self, _agent_id: usize) -> usize { 0 }`) is non-breaking
- Mirrors the pattern of the existing adapter layers in this codebase

**Q3: Differentiable (learned) comms vs. fixed protocol?**

Recommend **phased**:
- Phase 1: Fixed discrete vocab (extend action dims) — immediately achievable with `MultiDiscreteMlpBurnPolicy`
- Phase 2: Differentiable relaxation (Gumbel-softmax over message logits) — requires continuous policy support; not yet available in thrust
- Design note should document the Gumbel-softmax path and leave hooks for it

**Q4: `crossbeam-channel` for runtime plumbing?**

For Phase 1/2 (synchronous step-level comms): no new channels needed. The env's mutable state is the message bus within `step_multi`. For Phase 3 (async, deferred delivery, partial observability): `crossbeam-channel` is the right primitive and is already available under the `training` feature. The design note should call this boundary explicitly.

---

## Recommended Design Note Outline (`docs/COMMS_DESIGN.md`)

Follow the convention of `docs/MULTI_AGENT_DESIGN.md` (top-level uppercase filename, no date prefix):

```
# Multi-Agent Communication Channels — Design Note

## Current State
- BB env embedded comms (action-space model, proven but non-general)
- messages.rs infra (trainer-layer, not env-layer)
- trait gap

## Recommended Comms Model
1. Action-space extension for discrete fixed-vocab (Phase 1)
2. CommunicatingEnvironment supertrait (Phase 2)
3. Differentiable comms hooks (Phase 3, deferred)

## API Sketch
- CommunicatingEnvironment trait definition
- Updated MultiAgentResult (or message routing in step_multi)
- AgentMessage type (src/multi_agent/messages.rs addition)

## Design Questions Answered (all four with rationale)

## Phase Breakdown
- Phase 1: Trait + reference env (2–3 days)
- Phase 2: Joint trainer integration (1–2 days)
- Phase 3: Differentiable comms (deferred, needs continuous policy first)

## Out of Scope
- crossbeam channels for in-episode comms (Phase 1/2)
- Async partial-observability delivery (Phase 3+)
- Continuous action policy support (prerequisite to Phase 3)
```

---

## Acceptance Criteria

This epic is done when:

- [ ] `docs/COMMS_DESIGN.md` exists on `main` and answers all four design questions with a recommended approach and rationale
- [ ] The design note includes an API sketch (trait definition, type signatures) for `CommunicatingEnvironment` supertrait and any `AgentMessage` type
- [ ] The design note explicitly scopes what is in/out for Phase 1 vs Phase 2 vs Phase 3
- [ ] At least 3 child implementation issues are filed with `loom:triage`, one per phase, each linking back to #266
- [ ] The child issues cover: (a) `CommunicatingEnvironment` trait + reference env, (b) joint-trainer integration and comms loss term hook, (c) differentiable comms / Gumbel-softmax path (stretch, can be marked explicitly as deferred)

---

## Test Plan

This epic's deliverable is a design document + child issue decomposition; verification is review-based:

- [ ] Manual review: `docs/COMMS_DESIGN.md` exists and is coherent prose (not placeholder)
- [ ] All four design questions are addressed with rationale (not just listed)
- [ ] API sketches are valid Rust trait syntax (compile-check not required at design stage, but should be plausible)
- [ ] Child issues filed with `loom:triage` and cross-linked to #266
- [ ] No automated tests required for the design note itself; each child implementation issue will carry its own test plan

---

## Affected Files

Reference files (must exist on `origin/main` for builder to read — verified):

- `src/multi_agent/environment.rs` — current `MultiAgentEnvironment` trait (primary reference)
- `src/multi_agent/messages.rs` — existing `AgentMessage`-adjacent types to extend
- `src/multi_agent/joint.rs` — Slepian-Wolf joint trainer pattern
- `src/env/games/bucket_brigade/env.rs` — BB embedded-comms pattern (action-space model)
- `docs/MULTI_AGENT_DESIGN.md` — design note format convention to follow
- `Cargo.toml` — `crossbeam-channel` dep (verify scope/feature gate)

**Deliverable (new file to create):**

- `docs/COMMS_DESIGN.md` — does not exist yet; this is the primary output

**Child issues to create (no existing files):**
- 3 x `loom:triage` issues, one per phase

---

## Scope Guard

The builder implementing this epic must NOT:
- Modify `src/multi_agent/environment.rs` or any `.rs` source (that is Phase 1 implementation work)
- Begin any implementation of the comms channel
- Close this epic issue

The builder's work is complete when the design note is merged and the child issues are filed.

**Comment by rjwalters** [comment]
**Phase 1–2 experimental verdict (issue #304, PR #308):**

The SignalingGame + joint-PPO comms stack was run end-to-end (6 arms × 3 seeds × 500 iters; artifact + writeup in `docs/research/2026-07-signaling-protocol-emergence.md`). Verdict: **partial protocol emergence, two diagnosed bottlenecks**.

- **Shipped single-shot surface: no protocol possible, structurally.** `SignalingGame::step_multi` terminates every step and `collect_rollout`'s reset erases the message before the listener observes it — comms-on and comms-ablated arms are bit-identical at the chance floor (0.248 vs 1/V = 0.25).
- **With message persistence** (harness wrapper over the unmodified env, `episode_len = 8`): a load-bearing protocol partially emerges — reward 0.677 ± 0.172 vs 0.246 ± 0.008 ablated; end-to-end I(referent;guess) = 1.22 of 2.00 bits; best seed near-perfect (0.885, 1.88 bits), worst partial (0.464, 0.58 bits).
- **The Phase 2 comms-loss hook cannot shape the protocol**: the `comms_coef` {0, 0.01, 0.1} sweep yielded bit-identical policies (the penalty is a gradient-free constant by documented design).

Phases 1–2 infrastructure is functionally validated (message routing, heterogeneous per-agent obs dims, comms-coef plumbing all worked under real training load). The Phase 3 gate evidence is posted on #276.
